### ISSUE TITLE
Istio default retry strategy is transparently retrying POST requests that connected fine

### ISSUE THREAD

**Issue by Stono** [istio/istio#13851]
As you can see from this trace, a POST request reached vehicle-metric-service, and for some unknown reason (likely https://github.com/istio/istio/issues/13848) it 503's mid processing.

It was then retried by the default retry policy.

I *really* do not think the default retry policy should retry anything which successfully connected, only connection errors.  @nmittler I know we've talked about this before but couldn't find the associated thread.

<img width="1680" alt="Screenshot 2019-05-06 at 21 48 50" src="https://user-images.githubusercontent.com/1486729/57254327-e7c64e00-7048-11e9-9a4a-9577b0dae37b.png">

**Comment by Stono** [comment]
If this is to be expected then at the least you need to call it our or give people the ability to disable it, there are most definitely code paths we would not want to just retry without the source service being involved in that decision.

**Comment by nmittler** [comment]
So the issue here is that we include retries on 503 by default (which you can modify).

Envoy doesn't currently allow the configuration of retries on a per-method basis, but it probably should.

@PiotrSikora WDYT?

**Comment by Stono** [comment]
So the issue here is that without 503 retries things like https://github.com/istio/istio/issues/13848 become an even bigger problem.

The retries are currently masking that issue.

I can't help but think the 503 here too was completely unjust and linked to that issue.

**Comment by duderino** [comment]
@Stono servers are allowed to terminate their end of the connection during the idle period.  That should be considered routine behavior, but the client must handle that event well and I'm not sure it is.

**Comment by duderino** [comment]
@arianmotamedi if we had the Envoy PR that detects the closed by upstream you mentioned in https://github.com/istio/istio/issues/13848#issuecomment-489751037, we could retry 503s less aggressively which could address this issue.

@PiotrSikora @jplevyak @silentdai

**Comment by Stono** [comment]
@duderino - is there anything tracing what we talked about re separating 502 and 503, and only retrying 502/network level errors as that would address this issue I think

**Comment by ewolff** [comment]
https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html says POST is not idempotent. Shouldn't the default therefore be to not retry? Am I missing something?

**Comment by Stono** [comment]
Nope you're correct.  A POST imo shouldn't be retried unless the error was a connection error

**Comment by ewolff** [comment]
I am not sure I agree. The error could be raised after the action of the POST was performed. In that case, the retry would lead to an inconsistent state. https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5 doesn't really indicate otherwise.

**Comment by Stono** [comment]
Sorry perhaps I'm not being specific enough.  I mean if one envoy fails to connect to the other.

**Comment by ewolff** [comment]
That makes sense but then you would obviously need to ensure that the HTTP status was raised by the proxy and not the server. Also the proxy would need to ensure it did not forward the POST. I am not sure how reliable that can be done but I am certainly no expert.
Still I am not sure what the default should be. Maybe it should be optional?

**Comment by Stono** [comment]
Aye I completely agree with you and that's exactly the conversation I had with the istio folks a week or two ago.

At the moment 503 is used and covers too many scenarios.  Both envoy and upstream get wrapped by it

**Comment by ewolff** [comment]
Great, thanks for clarifying!

**Comment by arianmotamedi** [comment]
Here's my two cents on this: network errors happen. With Envoy or without, at some point something in the lower level causes the packets to get lost. When that happens, the client will almost never know if the request had reached the server for processing or not, so it has no choice but to either retry or report a failure. And as correctly mentioned here, http retries can have significant consequences for non-idempotent requests (POST, PUT, DELETE, etc.). Sure you could implement logic in Envoy to only retry on GET requests, but what about everything else? A 503 is simply "something happened up there and I didn't get what I expected back". A 503 won't include the state of the server, so there is **absolutely no way for Envoy (or ANY client) to 100% detect if the request is eligible to be retried or not** (only exception is a CONNECT failure which can be safely retried, but that's hardly ever the case for 503 errors). I want to emphasize on this because there seems to a misconception that a specific 503 flag (UC) means the issue is from Envoy-to-Envoy communication and it never made it to server. That's simply not true.

The real solution is to write code in a way that all requests (GET and POST) are idempotent from the server's perspective - that way failed requests can be safely retried. There are lots of articles out there on how to implement idempotency in microservices, but the simplest way is to have the client generate a unique id and pass it as a header to the server ("x-request-id"). Once the server gets the request, it processes it and stores the result for that id in a cache table. If a network-level error happens at this point, the client (Envoy) retries the request with the same id. This time the server sees that the request has already been processed, so it simply returns the cached result without processing the request again. Similarly if the connection is dropped before the request makes it to the server, next retry will cause the server to process it because it'll get a cache miss for that specific id.

I understand that this might not be the answer you were looking for, but unfortunately you can't have 100% resiliency by just relying on Envoy to do retries on its own.

**Comment by Stono** [comment]
I agree entirely.  Connection errors should be the only thing retried as I said in my OP, this retry happened on a POST after a successful connection.

So to clarify... the current strategy of retrying on 503 is wrong.

**Comment by Stono** [comment]
![503](https://user-images.githubusercontent.com/1486729/59775167-9f858700-92a8-11e9-8751-146006636028.png)

Just to further clarify what I meant by 503's masking connection errors, on this diagram there are 3 distinct connections, all of which can fail to connect.  At the moment, `C` which is a failed connection from destination envoy to the local app (lets say, because the app hasn't started), is manifested back as a `503` with a response flag of `UC`.  

This is what I mean about `503` being used to wrap too many scenarios, some of which likely can be retried (such as a connection issue between and envoy and an app).

**Comment by arianmotamedi** [comment]
If "C" fails with CONNECT_FAILURE, it needs to be retried by Dest Envoy, not Source. The only case Source envoy can retry is also on a CONNECT_FAILURE to Dest Envoy; any other 503 is ineligible unless idempotency is implemented in Dest app.

**Comment by Stono** [comment]
Yes I agree... I was simply telling you how it works **now**.  I really feel like we're saying the same thing here and I'm not sure how else to word it:  "Connection errors should be retried anywhere they can occur by the system instantiating the connection".

At the moment, they don't, they're wrapped in a 503 and retried by a default retry policy on all 503a.

**Comment by arianmotamedi** [comment]
Agreed, but I'd like to make an amendment to your statement: 

> Connection errors should be retried anywhere they can occur by the system instantiating the connection if and only if upstream can guarantee idempotency.

And I want to clarify - when I see the word "connection", I tend to think about **established** connections. When I say only CONNECT errors are always safe to retry, I mean the literal http CONENCT method to initiate a new connection. Once the connection is established (ACK'd), nothing else can be retried without knowing the implications.

**Comment by Stono** [comment]
Fair enough, I definitely mean un-acked connections, eg connections that have not been established.

**Comment by silentdai** [comment]
I personally agree that without deeper understanding of the 503 reason, istio shouldn't retry POST.
@mandarjog  Do you have more contexts on why istio still retry since #13848 is resolved.

**Comment by Stono** [comment]
You wanna have a chat with Josh who had a list of action items leading towards not retrying 503

**Comment by arianmotamedi** [comment]
@howardjohn @duderino as discussed, 503 errors should not be retried by default. It can be manually added if needed (and if the consequences are understood by the consumer). Unless there are any objections, I'm going to open up a pr to remove it from the default policy.

**Comment by Stono** [comment]
I agree with that move, however prepare yourself for a raft of "503" issues that get opened.  Those retries were added during a phase of istio getting a lot of stick about 503's, and right now they do retry 503s as a result of CONNECT errors too.    It might be worth pulling together some doco that explains how to debug 503s (similar to my blog post https://karlstoney.com/2019/05/31/istio-503s-ucs-and-tcp-fun-times/index.html) 

Also; i think there should be a broader "epic" covering the work @duderino captured some weeks ago (eg retries on CONNECT errors), not using 503 for everything, etc.

**Comment by arianmotamedi** [comment]
CONNECT errors aren't included in 503s, Envoy treats them separately. Istio's default retry policy includes both `connect-failure` as well as 503 as a retry-able status code: https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/retry/retry.go#L33. Removing 503 won't affect `connect-failure` retries.

**Comment by Stono** [comment]
OK if you're sure.  I just remember a lot of things like 503's during app rollouts because envoys had stale config (whilst waiting for pilot pushes) etc... 

In summary that retry was added for a reason, it's worth just double checking the underlying causes that can be addressed have been otherwise you'll be faced with all the old 503 drama again.

**Comment by Stono** [comment]
Hey, I'm rekindling this subject.  

Transparent retries of non-idempotent (POST) requests on anything other than gateway connection errors still doesn't feel right, so it'd be nice to get some traction on this discussion again.

**Comment by howardjohn** [comment]
@duderino @PiotrSikora @lambdai any update on this?

**Comment by sneko** [comment]
Hi, any news?

**Comment by duderino** [comment]
@yxue could you take a look at the retry actions in https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/retry/retry.go#L33 and see if you can find a combination that does not retry POSTs after the request has been sent (but does retry POSTs if the connection can't be made in the first place)?

I'm not entirely sure Envoy handles this correctly.  First I'd look at the Istio config, but you may have to go down into Envoy's retry policies and make a few changes there to really get this right.

Also @jplevyak should be able to give some excellent advice here.  If you have any questions about what the proper behavior is, please ask him.

**Comment by tariq1890** [comment]
@Stono Why did you close this issue? This seems to be relevant even now. Envoy retrying 503's is causing a lot of issues for us

**Comment by zhaohuabing** [comment]
We ran into this problem too.

**Comment by samucamarques** [comment]
The same problem here. Istio 1.7.4.
The microservices has its own mechanisms of transaction compensation on integration failures.
But Istio still retrying requests, when some integrations has 503 status code responses.
I need to disabled it. At least on POST. And let the application take care of it.

**Comment by donchev7** [comment]
Why not something like this to disable it?

`
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: "httpbin-virtual-service"
spec:
  hosts:
  - "*"
  gateways:
  - my-istio-gateway
  http:
  - match:
    - method:
        exact: POST
      uri:
        prefix: /status/503
    route:
    - destination:
        host: "httpbin"
        port:
          number: 80
    retries:
      attempts: 0
`

**Comment by defat** [comment]
Run into the same issue. Explicit retries disabling in VirtualService did help.

**Comment by brealsre** [comment]
Running into the same issue.  Has anyone come up with a work around?  Explicitly disabling the retries hasn't fixed it for us.   We are using Red Hat OpenShift Service Mesh which is using upstream 1.9.9

**Comment by dldiehl77** [comment]
We ran into the same issue too.  I experimented with changing our microservices to return 508 - Loop Detected instead of a 503 and this fixed it so that istio is not performing retries.  I suspect that many other values in the 5xx range would do the same trick, not just 508.  But, I'm not happy at the prospect of having to make a ton of code changes just to make it so that POSTs are not retried when the application returns a 503.  I agree, give us the option to let the application handle retries by being able to turn this "feature" off.
